### DISCUSSION TITLE
Compiling IM from source for ARM MacOS

### DISCUSSION THREAD

**Discussion by exekutive** [ImageMagick/ImageMagick#6446] [Help] [answered]
### ImageMagick version

7.1.1

### Operating system, version and so on

Ventura 13.5 on a M2 MBP

### Description

Homebrew installs very crippled versions of IM 
You don't get things like webpmux and the display module. Any time you convert an image to webp, metadata gets stripped.

Of course HB no longer supports build options.

Have people successfully compiled IM 7.1.1 on an ARM Mac from source?

**Comment by fmw42** [comment]
Should you not be complaining to Homebrew!

**Comment by exekutive** [comment]
No.  That would be a waste of time. I've given up on HB.
I'm not interested in complaining. I'm interested in solutions.

**Reply by fmw42** [reply]
Imagemagick has no control over how Homebrew makes its distributions.  They are the ones that I would think you need to notify and ask for a solution.  

Perhaps someone on the IM forum might know what to suggest. But I think you will do better on a Homebrew forum or asking them directly.

**Reply by exekutive** [reply]
maybe I didn't explain clearly enough. It is not possible with Homebrew's distribution. It likely never will be. A solution would involve using either a custom 3rd party formula, or compiling from source. My question is directed at those who have attempted or accomplished this. I'm wondering a) is it even possible without days of troubleshooting, and b) if it is, is there anything I should know?

**Reply by exekutive** [reply]
I completely understand that it's not within developer control or responsibility. But this is the discussion section right? I can ask other Mac users here?

**Reply by fmw42** [reply]
Yes, you are correct. Others may have an answer.  But out of curiosity, why won't Homebrew fix this on their end?

If you want another solution, not quite as simple, what I do is install all my delegates from MacPorts and then install Imagemagick from source. (MacPorts is just as bad, they only support IM 6 and refuse to have an alternate IM 7 version available).  I am on Mac OSX Ventura on M1 Mac mini.

**Reply by exekutive** [reply]
I thought I explained it already. Homebrew have permanently disabled build options for all distributions, and the core formula doesn't include webpmux or delegates needed to make 'display' work

**Comment by exekutive** [comment]
As I was afraid of, I ran into problems compiling and got stuck. Nothing is ever easy.

I downloaded the 7.1.1 source. The IM webpage for installing from source has instructions (https://imagemagick.org/script/install-source.php)  for Linux and Windows, but not Mac. For Mac they suggest HB, or they provide an **Intel** binary. Nonetheless I dove in. An 'Install-mac.txt' file is provided in the package which is nice, but it's probably outdated. Of course it recommends using HB to install, which is useless if you want even core functionality such as 'display'.  For compiling from source, it recommends using Macports, which I do not feel like installing (I hear MacPorts conflicts with HB and I don't need another headache). But, I already had the suggested delegates installed from before anyway (freetype, librsvg, graphviz) , so I went ahead.

I ran the config
`./configure --prefix=/opt --with-quantum-depth=16 --disable-dependency-tracking --without-perl`

it appeared to execute OK. It found the suggested delegates.

Then I ran make, and got stuck on an error for lqr

```
/Library/Developer/CommandLineTools/usr/bin/make  all-am
  CC       utilities/magick.o
  CC       MagickCore/libMagickCore_7_Q16HDRI_la-accelerate.lo
...
  CC       filters/MagickCore_libMagickCore_7_Q16HDRI_la-analyze.lo
  CCLD     MagickCore/libMagickCore-7.Q16HDRI.la
ld: library not found for -llqr-1
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [MagickCore/libMagickCore-7.Q16HDRI.la] Error 1
make: *** [all] Error 2
```

I already have liblqr installed so I don't know what its problem is.
This is madenning.

The Linux instructions also suggest

> For advanced users, we recommend a modules build:    $ ./configure --with-modules

So I tried that, but it stopped on an error as well.
`configure: error: libltdl is required for modules and OpenCL builds`
libltdl seems to be a component of 'libtool', which was already installed.
I'm guessing this is because HB crippled the libraries it installed, or this is just some Mac compatibility thing.

**Comment by exekutive** [comment]
I tried adding the` --with-lqr=no` flag to configure to skip lqr support for now, but IM still won't compile:

```
/Library/Developer/CommandLineTools/usr/bin/make  all-am
  CCLD     MagickCore/libMagickCore-7.Q16HDRI.la
Undefined symbols for architecture arm64:
  "_lqr_carver_destroy", referenced from:
      _LiquidRescaleImage in libMagickCore_7_Q16HDRI_la-resize.o
  "_lqr_carver_get_height", referenced from:
      _LiquidRescaleImage in libMagickCore_7_Q16HDRI_la-resize.o
  "_lqr_carver_get_width", referenced from:
      _LiquidRescaleImage in libMagickCore_7_Q16HDRI_la-resize.o
  "_lqr_carver_init", referenced from:
      _LiquidRescaleImage in libMagickCore_7_Q16HDRI_la-resize.o
  "_lqr_carver_new_ext", referenced from:
      _LiquidRescaleImage in libMagickCore_7_Q16HDRI_la-resize.o
  "_lqr_carver_resize", referenced from:
      _LiquidRescaleImage in libMagickCore_7_Q16HDRI_la-resize.o
  "_lqr_carver_scan_ext", referenced from:
      _LiquidRescaleImage in libMagickCore_7_Q16HDRI_la-resize.o
  "_lqr_carver_scan_reset", referenced from:
      _LiquidRescaleImage in libMagickCore_7_Q16HDRI_la-resize.o
  "_lqr_carver_set_preserve_input_image", referenced from:
      _LiquidRescaleImage in libMagickCore_7_Q16HDRI_la-resize.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [MagickCore/libMagickCore-7.Q16HDRI.la] Error 1
make: *** [all] Error 2
```

**Reply by dlemstra** [reply]
You should use `--without-lqr` instead.

**Reply by exekutive** [reply]
Thanks. I tried that and got the exact same result.

According to the install document:
`--without-something is equivalent to  --with-something=no`

**Comment by exekutive** [comment]
I'm on the right track ...
I manually compiled/installed lqr and libtool
(I suspect there are issues stemming from HB changing the default install directory for packages from usr to opt, and now I have a buttload of dead sym links. Is there anything HB didn't bork up??)

Anyway then I did make a clean. Then
```
./configure --prefix=/opt/ImageMagick-7.1.1 --with-modules
sudo make
sudo make install
```
no compile errors any more

x11 display module works
webp works properly and metadata is being preserved

however convert is not decoding jpegs, so I have more fixing to do.

**Comment by exekutive** [comment]
OK so jpeg is installed but IM can't find the header file
config log: `fatal error: 'jpeglib.h' file not found`
The file definitely exists, in `/opt/homebrew/Cellar/jpeg/9e/include/jpeglib.h`

tried to reinstall jpeg and got this:
```
jpeg is keg-only, which means it was not symlinked into /opt/homebrew,
because it conflicts with `jpeg-turbo`
```
There is a similar discussion about it here:
https://trac.macports.org/ticket/38907
jpeg-turbo is hardware accelerated jpeg, which I want. It also has a header file ....
`/opt/homebrew/Cellar/jpeg-turbo/3.0.0/include/jpeglib.h`

**Comment by exekutive** [comment]
OK finally got jpeg working as well. I had to specify everything explicitely:
`./configure CPPFLAGS=-I/opt/homebrew/Cellar/jpeg-turbo/3.0.0/include LDFLAGS=-L/opt/homebrew/Cellar/jpeg-turbo/3.0.0/lib --prefix=/opt/ImageMagick-7.1.1 --with-modules --with-jpeg=yes`

Now let's see what else is missing ....

**Comment by exekutive** [comment, marked as answer]
I solved most of my problems by reinstalling Apple command line tools and homebrew
